src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace.cs

```diff
@@ -295,31 +292,41 @@ private static bool HasProjectFileChanges(ProjectChanges changes)
                    changes.GetRemovedAnalyzerReferences().Any();
         }
 
+        private BuildHostProcessManager? _applyChangesBuildHostProcessManager;
 
-        private IProjectFile? _applyChangesProjectFile;
-#endif
+        /// <summary>
+        /// The loaded project file that we are currently applying changes to. This is set in <see cref="ApplyProjectChanges(ProjectChanges)"/> if we're modifying a project that is going to require
+        /// file changes; it's cleared once we're done with that project.
+        /// </summary>
+        private IRemoteProjectFile? _applyChangesProjectFile;
 
         public override bool TryApplyChanges(Solution newSolution)
         {
-#if SUPPORT_TRYAPPLYCHANGES
             return TryApplyChanges(newSolution, CodeAnalysisProgress.None);
-#else
-            return base.TryApplyChanges(newSolution);
-#endif
         }
 
-#if SUPPORT_TRYAPPLYCHANGES
-
         internal override bool TryApplyChanges(Solution newSolution, IProgress<CodeAnalysisProgress> progressTracker)
         {
             using (_serializationLock.DisposableWait())
             {
-                return base.TryApplyChanges(newSolution, progressTracker);
+                try
+                {
+                    Debug.Assert(_applyChangesBuildHostProcessManager == null);
+                    _applyChangesBuildHostProcessManager = new BuildHostProcessManager();
+
+                    return base.TryApplyChanges(newSolution, progressTracker);
+                }
+                finally
+                {
+                    _applyChangesBuildHostProcessManager!.DisposeAsync().AsTask().Wait();
```

Comment:
wow :'(

my hate is strong here :)

Reply:
do we need to do this "create" and "synchronously dispose and clear" pattern?  could we instead just make a single manager that stays alive which we know we're serializging messages to by virtue of hte lock above?

Reply:
@CyrusNajmabadi not sure I followed this comment -- we _are_ making a single manager that stays alive; we just want to  ensure the processes are shut down once we're done. I could also just leave them around for forever but wasn't sure if that'd create other surprises -- right now we're making this API launch processes when it didn't before, leaving it running might make further surprises.